### PR TITLE
Make `find_similar_impl_candidates` a little fuzzier.

### LINKED ISSUE #92113: Regression in diagnostic quality for unimplemented traits on arrays.
Commit dcd716fee2303c93a0bf1ba68e667c9bc4bf3558 (https://github.com/rust-lang/rust/pull/86986) appears to have regressed the quality of compiler diagnostics produced when an array type does not satisfy a trait but its slice version does (there may be other scenarios that see similar regressions, but this is where I've seen it).

This is currently blocking https://github.com/rust-lang/rust/pull/91314.

Motivating example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c9f9bb3a828c28cc18c6591d545fb5db)):

```rust
trait Read {}

impl Read for &[u8] {}

fn wants_read(_: impl Read) {}

fn main() {
    wants_read([0u8]);
}
```

When compiled on dcd716fee2303c93a0bf1ba68e667c9bc4bf3558~1 (83b32f27fc6c34b0b411f47be31ab4ae07eafed4), rustc provides a help message telling the user that the trait is implemented for slices:

```
error[E0277]: the trait bound `[u8; 1]: Read` is not satisfied
 --> ../scratch/issue-90528-unsizing-suggestion-1.rs:8:16
  |
8 |     wants_read([0u8]);
  |     ---------- ^^^^^ the trait `Read` is not implemented for `[u8; 1]`
  |     |
  |     required by a bound introduced by this call
  |
  = help: the following implementations were found:
            <&[u8] as Read>
note: required by a bound in `wants_read`
 --> ../scratch/issue-90528-unsizing-suggestion-1.rs:5:23
  |
5 | fn wants_read(_: impl Read) {}
  |                       ^^^^ required by this bound in `wants_read`
```

Compiled on dcd716fee2303c93a0bf1ba68e667c9bc4bf3558, this help message no longer appears:

```
error[E0277]: the trait bound `[u8; 1]: Read` is not satisfied
 --> ../scratch/issue-90528-unsizing-suggestion-1.rs:8:16
  |
8 |     wants_read([0u8]);
  |     ---------- ^^^^^ the trait `Read` is not implemented for `[u8; 1]`
  |     |
  |     required by a bound introduced by this call
  |
note: required by a bound in `wants_read`
 --> ../scratch/issue-90528-unsizing-suggestion-1.rs:5:23
  |
5 | fn wants_read(_: impl Read) {}
  |                       ^^^^ required by this bound in `wants_read`
```

Inspiration for regression tests can be taken from https://github.com/rust-lang/rust/pull/91314.

@rustbot modify labels: +A-diagnostics +T-compiler +D-terse

cc @lcnr

### PR DESCRIPTION
**Pull request by BGR360**
Resolves #92113 

This is the initial draft, I'm curious for your thoughts. Still need to write ui tests for the arrays/slices case.

Let me know if anything is unclear; I stayed up far too late and am far too tired to document my decisions in more detail at this very moment 😁 

r? @lcnr

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/on_unimplemented.rs [hunk 1]
```diff
@@ -56,7 +57,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                     trait_ref.substs.types().skip(1),
                     impl_trait_ref.substs.types().skip(1),
                 )
-                .all(|(u, v)| self.fuzzy_match_tys(u, v))
+                .all(|(u, v)| self.fuzzy_match_tys(u, v, StripReferences::No))
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/on_unimplemented.rs:60]: are there unwanted changes when using `StripReferences::Yes` here?

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -1458,40 +1490,75 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
...
+                                    trait_ref: imp,
+                                    similarity: CandidateSimilarity::Simplified,
+                                });
+                            }
+                        }
+
+                        // Check for fuzzy match.
+                        // Pass `StripReferences::Yes` because although we do want to
+                        // be fuzzier than `simplify_type`, we don't want to be
+                        // *too* fuzzy.
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1529]: that reasoning seems unclear to me :thinking: how does passing `StripReferences::Yes` prevent ourselves from being "*too* fuzzy"?

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 3]
```diff
@@ -1458,40 +1490,75 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
...
+                            StripReferences::Yes,
+                        );
+                        if let Some(imp_simp) = imp_simp {
+                            if simp == imp_simp {
+                                return Some(ImplCandidate {
+                                    trait_ref: imp,
+                                    similarity: CandidateSimilarity::Simplified,
+                                });
+                            }
+                        }
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:1524]: i don't believe that this special case has a meaningful advantage over only using exact matches and fuzzy matches.
I would prefer removing this and removing the `StripReferences` function parameter from `simplify_type`. This should be the only place where it is currently used.

**Review by lcnr — COMMENTED** [review]
it might be meaningful to also sort by whether the non-self parameters of the impls also match the searched ones.

would be nice to try this here if you're interested, otherwise we can also merge this as is.

i might not be available over the holidays, so you may have to wait a week until i can come back to you

**Comment by JohnCSimon** [comment]
Ping from triage: 
@BGR360 can you post your status on this PR? Thanks.

**Comment by BGR360** [comment]
Sorry, been a lot of other stuff on my mind recently. I don't have immediate plans to continue on this, unless someone really wants this to get in.

**Comment by lcnr** [comment]
@BGR360 went and continued this PR in #93298, hope I understood your comment correctly and you don't mind '^^

**Comment by BGR360** [comment]
@lcnr I don't mind at all, thank you for carrying the torch for me!

**Comment by lcnr** [comment]
closing it as i expect us to merge #93298, again thank you @BGR360

**Pull request closed without merging**